research/condist-fl/run_validate.py

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from argparse import ArgumentParser
+from pathlib import Path, PurePath
+from typing import Dict
+
+import torch
+import numpy as np
+import pandas as pd
+from src.data import DataManager
+from src.validator import Validator
+from src.utils.get_model import get_model
+
+
+def load_ckpt(app: str, model: torch.nn.Module, ckpt_path: str):
+    ckpt = torch.load(ckpt_path)
+    if app == "app_server":
+        state_dict = ckpt["model"]
+    else:
+        state_dict = ckpt["state_dict"]
+    model.load_state_dict(state_dict)
+    return model
+
+
+def run_validation(args):
+    ws = Path(args.workspace)
+    if not ws.exists():
+        raise ValueError(f"Workspace path {ws} does not exists.")
+    prefix = ws / "simulate_job"
+
+    server = "app_server"
+    clients = [p.name for p in prefix.glob("app_*") if "server" not in p.name]
+
+    # Collect all checkpoints to evaluate
+    checkpoints = {
+        "app_server": str(prefix / "app_server/best_FL_global_model.pt")
+    }
+    checkpoints.update({
+        c: str(prefix / c / "models/best_model.ckpt") for c in clients
+    })
+
+    # Collect configs from clients
+    config = {
+        c: {
+            "data": str(prefix / c / "config/config_data.json"),
+            "task": str(prefix / c / "config/config_task.json")
+        } for c in clients
+    }
+
+    metrics = {app: {} for app in [server] + clients}
+    for client in clients:
+        with open(config[client]["data"]) as f:
+            data_config = json.load(f)
+
+        with open(config[client]["task"]) as f:
+            task_config = json.load(f)
+
+        # The first argument `app_root` seems to be useless
+        dm = DataManager(str(prefix), data_config)
+        dm.setup("test")
+
+        # Create model & validator from task config
+        model = get_model(task_config["model"])
+        validator = Validator(task_config)
+
+        # Run server validation
+        model = load_ckpt("app_server", model, checkpoints["app_server"])
+        model = model.eval().cuda()
+        raw_metrics = validator.run(model, dm.get_data_loader("test"))
+        raw_metrics.pop("val_meandice")
+        metrics["app_server"].update(raw_metrics)
+
+        # Run client validation
+        for c in clients:
+            model = load_ckpt(c, model, checkpoints[c])
```

Comment:
Can you add some printout to show which model is evaluated on which data? Currently, the output just shows this
```
root@5037762:/workspace_hroth/Code/nvflare/condist-fl/research/condist-fl# python run_validate.py -w workspace -o cross_site_validate.json
Loading dataset: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 56/56 [02:02<00:00,  2.19s/it]
Validation DataLoader: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 56/56 [02:09<00:00,  2.32s/it]
Validation DataLoader: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 56/56 [02:09<00:00,  2.32s/it]
Validation DataLoader: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 56/56 [01:03<00:00,  1.13s/it]
Validation DataLoader: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 56/56 [01:03<00:00,  1.14s/it]
Validation DataLoader: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 56/56 [01:03<00:00,  1.13s/it]
Loading dataset: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 26/26 [03:56<00:00,  9.10s/it]
Validation DataLoader: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 26/26 [00:51<00:00,  1.96s/it]
Validation DataLoader: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 26/26 [00:51<00:00,  1.96s/it]
Validation DataLoader: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 26/26 [00:51<00:00,  1.96s/it]
```

Reply:
Added print functions to identify which dataset is loaded and which model is using for validation.